#590 Added Fork chain implementation and tests

CarlosGines opened the pull request:
Fixes for #590:
- Added `FkChain` class.
- Added corresponding unit tests.
- Simplified `TkFork` implementation by the use of `FkChain`.

davvd commented:
@carlosgines Thanks, I will ask someone to do a code review here, thanks for your pull request

davvd commented:
@mkordas please review,thanks

mkordas commented:
@CarlosGines I'll review that one soon

mkordas commented on src/test/java/org/takes/facets/fork/FkChainTest.java:
```diff
@@ -0,0 +1,68 @@
+     * @throws IOException If some problem inside
+     */
+    @Test
+    public void dispatchesByRegularExpression() throws IOException {
```
@CarlosGines please see https://github.com/teamed/qulice/issues/689

mkordas commented on src/main/java/org/takes/facets/fork/FkChain.java:
```diff
@@ -0,0 +1,83 @@
+ * a response or until none are left.
+ * @author Carlos Gines (efrel.v2@gmail.com)
+ * @version $Id$
+ * @since 0.32
```
@CarlosGines `0.32` is already released, please use next version

mkordas commented on src/main/java/org/takes/facets/fork/FkChain.java:
```diff
@@ -0,0 +1,83 @@
+     * @param forks Forks
+     */
+    public FkChain(final Collection<Fork> forks) {
+        this.forks = Collections.unmodifiableCollection(forks);
```
@CarlosGines making `Collections.unmodifiableCollection` is not enough here, `forks` can still be modified from outside

pinaf replied:
@mkordas please explain what you mean

mkordas replied:
@pinaf @CarlosGines see http://stackoverflow.com/questions/35071782/collections-unmodifiablecollection-in-constructor

CarlosGines replied:
So you mean we could improve protection making a copy of the list before wrapping, like this:
`this.forks = Collections.unmodifiableCollection(new ArrayList<>(forks));` 
Or using the Guava class mentioned on that thread. 
I see your point, but it is not done like that anywhere in the project. You can find many examples of `Collection.unmodifiableXXX` all over the project and it is always done the same way as here (that is why I did it). Do you want me to change it in spite of this?

mkordas replied:
@CarlosGines `unmodifiableCollection` is incorrect approach, but I'll submit separate issue for that, so let's leave as it is

mkordas commented on src/main/java/org/takes/facets/fork/FkChain.java:
```diff
@@ -0,0 +1,83 @@
+import org.takes.misc.Opt;
+
+/**
+ * A {@link Fork} chain. Routes to each {@link Fork} in order until one returns
```
@CarlosGines usually `@link` tags are not recommended in first sentence as they make it harder to read

mkordas commented:
@CarlosGines see minor comments, I'll check again here tomorrow

pinaf commented:
@CarlosGines let's add a test for the case where no forks match the request.

CarlosGines commented:
@mkordas @pinaf thank you. Here are the new fixes:
- Tests now throw `Exception`.
- Version updated in javadoc to `@version 0.33`.
- Removed `{@link Fork}` from the first line.
- Added test for the case where no forks match the request.

I would also need more explanations about why `Collections.unmodifiableCollection(forks)` is not enough.

mkordas commented on src/main/java/org/takes/facets/fork/FkChain.java:
```diff
@@ -0,0 +1,83 @@
+import org.takes.misc.Opt;
+
+/**
+ * A Fork chain. Routes to each given Fork in order until one returns
```
@CarlosGines I don't quite understand second sentence, especially part `in order until one`

pinaf replied:
@mkordas if you add a comma after `in order` it is easier to understand?

mkordas replied:
@pinaf yes, a bit more

mkordas commented on src/main/java/org/takes/facets/fork/FkChain.java:
```diff
@@ -0,0 +1,83 @@
+/**
+ * A Fork chain. Routes to each given Fork in order until one returns
+ * a response or until none are left.
+ * @author Carlos Gines (efrel.v2@gmail.com)
```
@CarlosGines shouldn't we add also @pinaf as an author here?

CarlosGines replied:
I don't know, tell me whether I need to change it to pass the review.

mkordas commented on src/test/java/org/takes/facets/fork/FkChainTest.java:
```diff
@@ -0,0 +1,82 @@
+                    new FkRegex("/g[a-z]{2}", ""),
+                    new FkRegex("/h[a-z]{2}", body),
+                    new FkRegex("/i[a-z]{2}", "")
+                ).route(new RqFake("GET", "/hey?yu")).get()
```
@CarlosGines I have problems understanding this test. How `/h[a-z]{2}` is supposed to match `/hey?yu`?

CarlosGines replied:
Well, it does match. Check here for example to get more info: https://regex101.com

mkordas replied:
@CarlosGines OK, it just matches partially

CarlosGines replied:
@mkordas partially it is enough, as long as we don't use the anchors `^` and `$`

mkordas commented on src/test/java/org/takes/facets/fork/FkChainTest.java:
```diff
@@ -0,0 +1,82 @@
+                ).route(new RqFake("GET", "/hey?yu")).get()
+            ).print(),
+            Matchers.equalTo(
+                Joiner.on("\r\n").join(
```
@CarlosGines `\r\n`? Isn't specific to Windows?

CarlosGines replied:
I am just using what I have seen in other working tests in the project. It seems to be done because of [this](https://github.com/yegor256/takes/blob/e0390168f34a0659ff1baa2a6c4ca92bd6d38294/src/main/java/org/takes/rs/RsPrint.java#L124).

mkordas replied:
@CarlosGines thanks, I see

mkordas commented:
@CarlosGines please see comments above

CarlosGines commented:
@mkordas new commit improving the sentence on `FkChain` javadoc to be more understandable. See my answers on the rest of comments.

CarlosGines commented:
@mkordas next time please:
- Give me all reviews at the same time.
- Be specific about what you want me to change in order to pass the review.

This way we will not need to come back to the same task several times and waste time.

CarlosGines commented:
@mkordas anything else that needs to be addressed? If there is nothing else, could you please confirm the review as done?

mkordas commented:
@CarlosGines sorry for partial reviews, I didn't have time to full review when I started, but I wanted to give early feedback. Other option was to start one day later.

mkordas commented:
@CarlosGines looks good now!

mkordas commented:
@rultor merge

rultor commented:
> @rultor merge

@mkordas Thanks for your request. @yegor256 Please confirm this.

CarlosGines commented:
@mkordas thank you very much!

yegor256 commented:
@rultor merge pls

rultor commented:
> @rultor merge pls

@yegor256 OK, I'll try to merge now. You can check the progress of the merge [here](http://www.rultor.com/t/7480-192533667)

rultor commented:
> @rultor merge pls

@yegor256 Done! FYI, the full log is [here](http://www.rultor.com/t/7480-192533667) (took me 11min)

davvd commented:
@elenavolokhova please, review this task for compliance with our [quality rules](http://at.teamed.io/qa.html)

davvd commented:
@rultor deploy now pls

rultor commented:
> @rultor deploy now pls

@davvd OK, I'll try to deploy now. You can check the progress [here](http://www.rultor.com/t/7480-192957516)

rultor commented:
> @rultor deploy now pls

@davvd Done! FYI, the full log is [here](http://www.rultor.com/t/7480-192957516) (took me 16min)

elenavolokhova commented:
@davvd Looks good!

davvd commented:
> @davvd Looks good!

@elenavolokhova thanks, got it, great to see high-quality tickets here

davvd commented:
@mkordas just added 10 mins to @elenavolokhova (for QA), payment ID is `79800242`. Many thanks! **34 mins** were added to your account in Transaction ID `56df272308baac00ad0001db` (task took 78 hours and 42 mins). extra minutes for review comments (c=19). +34 added to your rating, current score is: [+8471](http://www.netbout.com/b/36080?open=rating)
